File: formcraft-plm/src/main/java/fr/formcraft/model/entity/ChangeRequest.java
package fr.formcraft.model.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import fr.formcraft.model.enums.ChangeRequestStatus;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

/** Engineering/Formulation Change Request (ECR/ECO) against a product. */
@Entity
@Table(name = "change_requests")
@Getter
@Setter
@NoArgsConstructor
public class ChangeRequest {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    @JsonIgnoreProperties({"compositionLines", "nutrientValues", "formulationResults", "workflowTasks", "qualityChecks"})
    private Product product;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(columnDefinition = "TEXT")
    private String reason;

    @Column(columnDefinition = "TEXT")
    private String impact;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ChangeRequestStatus status = ChangeRequestStatus.DRAFT;

    @Column(name = "requested_by", length = 100)
    private String requestedBy;

    @Column(name = "requested_at", nullable = false, updatable = false)
    private LocalDateTime requestedAt = LocalDateTime.now();

    @Column(name = "decided_by", length = 100)
    private String decidedBy;

    @Column(name = "decided_at")
    private LocalDateTime decidedAt;

    @Column(name = "decision_comment", columnDefinition = "TEXT")
    private String decisionComment;
}

File: app/schemas/change_request.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.product import ProductBase
from app.schemas.base import CamelModel


class ChangeRequestSchema(CamelModel):
    id: int
    product: ProductBase
    title: str
    description: str | None = None
    reason: str | None = None
    impact: str | None = None
    status: str
    requested_by: str | None = None
    requested_at: datetime
    decided_by: str | None = None
    decided_at: datetime | None = None
    decision_comment: str | None = None


class ChangeRequestCreateRequest(BaseModel):
    productId: int
    title: str = Field(..., min_length=1)
    description: str | None = None
    reason: str | None = None
    impact: str | None = None


class DecisionRequest(BaseModel):
    status: str
    decisionComment: str | None = None
